subgraphs/playback/src/utils/server.ts

```diff
@@ -0,0 +1,181 @@
+import logger from '../logger';
+import * as Sentry from '@sentry/node';
+
+export const app = express();
+export const httpServer = http.createServer(app);
+
+let typeDefs = gql(
+  readFileSync('schema.graphql', {
+    encoding: 'utf-8',
+  })
+);
+const schema = buildSubgraphSchema({
+  typeDefs,
+  resolvers,
+});
+const wsServer = new WebSocketServer({
+  server: httpServer,
+  path: '/ws',
+});
+const serverCleanup = useServer(
+  {
+    schema,
+    onConnect: (ctx) => {
+      if (ctx.connectionParams?.['authorization']) return true;
+      if (ctx.extra.request.headers?.['authorization']) return true;
+
+      //If there is no authorization, we will mock everything and use the referer for the users identifier in local mocked data
+      return true;
+    },
+    context: (ctx) => {
+      const routerAuthorization =
+        (ctx.connectionParams?.['authorization'] as string) ??
+        (ctx.extra.request.headers?.['authorization'] as string) ??
+        '';
+      checkRouterSecret(routerAuthorization);
+      const token =
+        (ctx.connectionParams?.['authorization'] as string) ??
+        (ctx.extra.request.headers?.['authorization'] as string);
+
+      const defaultCountryCode =
+        (ctx.connectionParams?.['country-code'] as string) ??
+        (ctx.extra.request.headers?.['country-code'] as string) ??
+        'US';
+
+      if (!token) {
+        const userIdForMocks =
+          ctx.extra.request.rawHeaders['x-graphos-id'] ?? 'default';
+        addUser(userIdForMocks);
+
+        return {
+          defaultCountryCode,
+          dataSources: {
+            spotify: new MockedSpotifyDataSource(userIdForMocks),
+          },
+          userIdForMocks,
+        };
+      }
+
+      return {
+        defaultCountryCode,
+        dataSources: {
+          spotify: new SpotifyAPI({
+            cache: callbackApolloServer.cache,
+            token,
+          }),
+        },
+      };
+    },
+  },
+  wsServer
+);
+
+const sentryPlugin: ApolloServerPlugin<ContextValue> = {
+  async requestDidStart() {
+    return {
+      async didEncounterErrors(ctx) {
+        for (const err of ctx.errors) {
+          if (err.extensions?.code == 'GRAPHQL_VALIDATION_FAILED')
+            Sentry.withScope((scope) => {
+              // Annotate whether failing operation was query/mutation/subscription
+              scope.setTag('message', err.message);
+              // Log query and variables as extras
+              // (make sure to strip out sensitive data!)
+              scope.setExtra('query', ctx.request.query);
+              scope.setExtra('variables', ctx.request.variables);
+              if (err.path) {
+                // We can also add the path as breadcrumb
+                scope.addBreadcrumb({
+                  category: 'query-path',
+                  message: err.path.join(' > '),
+                  level: 'debug',
+                });
+              }
+              Sentry.captureException(err);
+            });
```

Comment:
```suggestion
          if (err.extensions?.code == 'GRAPHQL_VALIDATION_FAILED') {
            Sentry.withScope((scope) => {
              // Annotate whether failing operation was query/mutation/subscription
              scope.setTag('message', err.message);
              // Log query and variables as extras
              // (make sure to strip out sensitive data!)
              scope.setExtra('query', ctx.request.query);
              scope.setExtra('variables', ctx.request.variables);
              if (err.path) {
                // We can also add the path as breadcrumb
                scope.addBreadcrumb({
                  category: 'query-path',
                  message: err.path.join(' > '),
                  level: 'debug',
                });
              }
              Sentry.captureException(err);
            });
          }
```

For long blocks of code like this (actually... to be honest always) I prefer to wrap the if block in braces. Its much easier to distinguish between the body of the `if` and the rest of the code. It took me a few back-and-forths to realize the next batch of code was still part of the loop.

Reply:
Updated in [9ac8554](https://github.com/apollographql/spotify-showcase/pull/94/commits/9ac8554784f26dee7c5f80253db7fb7953b3a44a)